CalculatorEngine.cs

```diff
@@ -0,0 +1,131 @@
+﻿using System;
+using System.Linq;
+
+namespace GettingStartedWithCSharp
+{
+    public class CalculatorEngine : ICalculatorEngine
+    {
+        private decimal _value = 0m;
+        private decimal _memory;
+        public bool HasMemoryStored { get; private set; } = false;
+        private readonly IMessageBoxDisplayService _messageBoxDisplayService;
+        private readonly IUtils _utils;
+
+        public CalculatorEngine(IMessageBoxDisplayService messageBoxDisplayService, IUtils utils)
+        {
+            _messageBoxDisplayService = messageBoxDisplayService;
+            _utils = utils;
+        }
+
+        public decimal GetNumberToBeShown(string operation, decimal numarAfisat)
+        {
+            switch (operation)
+            {
+                case "+":
+                    _value = _value + numarAfisat;
+                    return _value;
+                case "-":
+                    _value =_value - numarAfisat;
+
+                    return _value;
+                case "*":
+                    _value = _value * numarAfisat;
+                    return _value;
+                case "/":
+                    try
+                    {
+                        _value = _value / numarAfisat;
+                    }
+                    catch (DivideByZeroException)
+                    {
+                        throw new ArgumentException(String.Format("Numarul {0} nu poate fi impartit la 0.", _value), "Value");
+                    }
+                    return _value;
+                case "sqrt":
+                    if (_value < 0)
+                    {
+                        throw new ArgumentException(String.Format("{0} este un numar negativ => Nu ii putem extrage radacina patratica!", _value), "Value");
+                    }
+
+                    _value = (decimal)(Math.Sqrt((double)_value));
+                    
+                    return _value;
+                default:
+                    return _value;
+            }
+        }
+
+        public void MemoryAdd(decimal newValue)
+        {
+            _memory += newValue;
+        }
+
+        public void MemoryDiff(decimal newValue)
+        {
+            _memory -= newValue;
+        }
+
+        public void MemoryStore(string textAfisat)
```

Comment:
Asta de ce are in continuare un parametru `string`? 

E treaba presenter-ului sa adapteze datele din display pentru clasele cu care colaboreaza.

Deci acest `decimal.Parse` este logica de Presenter care a ajuns sa polueze clasa de business logic.